src/lib/tuning/index/base_index_evaluator.cpp

```diff
@@ -0,0 +1,219 @@
+
+float BaseIndexEvaluator::_existing_memory_cost(const IndexChoice& index_choice) const {
+  auto table = StorageManager::get().get_table(index_choice.column_ref.table_name);
+  float memory_cost = 0.0f;
+  for (ChunkID chunk_id = ChunkID{0}; chunk_id < table->chunk_count(); ++chunk_id) {
+    auto chunk = table->get_chunk(chunk_id);
+    auto index = chunk->get_index(index_choice.type, index_choice.column_ref.column_ids);
+    if (index) {
+      memory_cost += index->memory_consumption();
+    }
+  }
+  return memory_cost;
+}
+
+void BaseIndexEvaluator::_inspect_query_cache() {
+  _access_records.clear();
+
+  // ToDo(anybody): The cache interface could be improved by introducing values() method in
+  // AbstractCache interfaceand implement in all subclasses
+  //  const auto& query_plan_cache = SQLQueryOperator::get_query_plan_cache().cache();
+  // -> so this implementation could be independent of the actual cache implementation
+
+  const auto& lqp_cache = SQLQueryCache<std::shared_ptr<AbstractLQPNode>>::get();
+  // We cannot use dynamic_pointer_cast here because SQLQueryCache.cache() returns a reference, not a pointer
+  auto gdfs_cache_ptr =
+      dynamic_cast<const GDFSCache<std::string, std::shared_ptr<AbstractLQPNode>>*>(&(lqp_cache.cache()));
+  if (!gdfs_cache_ptr) {
+    LOG_WARN("BaseIndexEvaluator can only analyze GDFSCache instances! Evaluations may be useless...");
+    return;
+  }
+
+  const auto& fibonacci_heap = gdfs_cache_ptr->queue();
+
+  LOG_DEBUG("Query plan cache (size: " << fibonacci_heap.size() << "):");
+  if (fibonacci_heap.size() == 0) {
+    LOG_WARN("There are no logical query plans in the cache. Make sure that logical query plans get cached!");
+  }
+
+  auto cache_iterator = fibonacci_heap.ordered_begin();
+  auto cache_end = fibonacci_heap.ordered_end();
+
+  for (; cache_iterator != cache_end; ++cache_iterator) {
+    const auto& entry = *cache_iterator;
+    LOG_DEBUG("  -> Query '" << entry.key << "' frequency: " << entry.frequency << " priority: " << entry.priority);
+    _inspect_lqp_operator(entry.value, entry.frequency);
+  }
+}
+
+void BaseIndexEvaluator::_inspect_lqp_operator(const std::shared_ptr<const AbstractLQPNode>& op,
+                                               size_t query_frequency) {
+  std::list<const std::shared_ptr<const AbstractLQPNode>> queue;
+  queue.push_back(op);
+  while (!queue.empty()) {
+    auto lqp_node = queue.front();
+    queue.pop_front();
+
+    if (auto left_child = lqp_node->left_child()) {
+      queue.push_back(left_child);
+    }
+    if (auto right_child = lqp_node->right_child()) {
+      queue.push_back(right_child);
+    }
+
+    switch (lqp_node->type()) {
+      case LQPNodeType::Predicate: {
+        auto predicate_node = std::dynamic_pointer_cast<const PredicateNode>(lqp_node);
+        DebugAssert(predicate_node, "LQP node is not actually a PredicateNode");
```

Comment:
Can this ever happen?

Reply:
I agree that this is unlikely to happen, but it could happen e.g. as result of a copy-paste error (implementing something similar to a PredicateNode). Since it is unlikely, we only check it during Debug. Still, as it could fail here, we wanted to give the person debugging it a lead with an explicit failure message, as opposed to an obscure crash.